daffodil-core/src/test/scala/org/apache/daffodil/infoset/TestInfosetFree.scala

```diff
@@ -0,0 +1,280 @@
+
+    // all elements freed, including hidden elements
+    val expectedXML =
+      <root freed="self" xmlns="http://example.com">
+        <e freed="self">1</e>
+      </root>
+
+    assertEquals(scala.xml.Utility.trim(expectedXML).toString, actualXML.toString)
+  }
+
+  @Test def testInfosetFree3(): Unit = {
+    val testSchema = SchemaUtils.dfdlTestSchema(
+      <xs:include schemaLocation="org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd" />,
+      <dfdl:format ref="tns:GeneralFormat" />,
+      <xs:element name="root">
+        <xs:complexType>
+          <xs:sequence>
+            <xs:element name="fieldLen" type="xs:int" dfdl:lengthKind="explicit" dfdl:length="1" />
+            <xs:element name="field" type="xs:int" maxOccurs="unbounded"
+              dfdl:lengthKind="explicit" dfdl:length="{ ../tns:fieldLen }" />
+          </xs:sequence>
+        </xs:complexType>
+      </xs:element>)
+
+    val actualXML = TestInfosetFree.test(testSchema, "1123".getBytes)
+
+    // all elements freed, execpted for fieldLen because it is used in an expression
+    val expectedXML =
+      <root freed="self" xmlns="http://example.com">
+        <fieldLen>1</fieldLen>
+        <field freed="self+array">1</field>
+        <field freed="self+array">2</field>
+        <field freed="self+array">3</field>
+      </root>
+
+    assertEquals(scala.xml.Utility.trim(expectedXML).toString, actualXML.toString)
+  }
+
+  @Test def testInfosetFree4(): Unit = {
+    val testSchema = SchemaUtils.dfdlTestSchema(
+      <xs:include schemaLocation="org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd" />,
+      <dfdl:format ref="tns:GeneralFormat" />,
+      <xs:element name="root">
+        <xs:complexType>
+          <xs:sequence>
+            <xs:element name="fieldCount" type="xs:int" dfdl:lengthKind="explicit" dfdl:length="1"
+              dfdl:outputValueCalc="{ fn:count(../tns:field) }" />
+            <xs:element name="field" type="xs:int" maxOccurs="unbounded"
+              dfdl:occursCountKind="expression" dfdl:occursCount="{ ../tns:fieldCount }"
+              dfdl:lengthKind="explicit" dfdl:length="1" />
+          </xs:sequence>
+        </xs:complexType>
+      </xs:element>)
+
+    val actualXML = TestInfosetFree.test(testSchema, "3123".getBytes)
+
+    // Most elements are not freed. Both fieldCount and field are used in an
+    // expression. Even though only the count of the field array is needed, and
+    // we could theoretically free the field elements but no the array, they
+    // are not currently freed. Our usedInAnExpression logic is not yet
+    // sophisticated enough to differentiate between how elements are used in
+    // an expression and how that affect releasability
+    val expectedXML =
+      <root freed="self" xmlns="http://example.com">
+        <fieldCount>3</fieldCount>
+        <field>1</field>
+        <field>2</field>
+        <field>3</field>
+      </root>
```

Comment:
This test is maybe more analagous to some more common formats, where we have lots of expression, and output value calc's that can make streaming difficult. In this particular case, nothing can be freed. In this case, we would need a feature that allows for compacting arrays and handling count of elements as different type of acces than element indicies.